src/NUnitEngine/nunit.engine/Services/TestSelectionParser.cs

```diff
@@ -0,0 +1,275 @@
+            }
+
+            if (terms.Count == 1)
+                return terms[0];
+
+            var sb = new StringBuilder("<or>");
+
+            foreach (string term in terms)
+                sb.Append(term);
+
+            sb.Append("</or>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single element or an and expression and return the xml
+        /// </summary>
+        public string ParseFilterTerm()
+        {
+            var elements = new List<string>();
+            elements.Add(ParseFilterElement());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != AND_OP1 && op != AND_OP2)
+                    break;
+
+                NextToken();
+                elements.Add(ParseFilterElement());
+            }
+
+            if (elements.Count == 1)
+                return elements[0];
+
+            var sb = new StringBuilder("<and>");
+
+            foreach (string element in elements)
+                sb.Append(element);
+
+            sb.Append("</and>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single filter element such as a category expression
+        /// and return the xml representation of the filter.
+        /// </summary>
+        public string ParseFilterElement()
+        {
+            Token nextOp = Peek();
+            if (nextOp == LPAREN || nextOp == NOT_OP)
+                return ParseExpressionInParentheses();
+
+            string lhs = GetWord();
+
+            switch (lhs)
+            {
+                case "id":
+                    string op = GetOperator("=", "==");
+                    string values = GetIdList();
+                    return string.Format("<id>{0}</id>", values);
+
+                case "cat":
+                case "category":
+                case "method":
+                case "class":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = GetWordList();
+                    string tag = lhs == "category" ? "cat" : lhs;
+                    switch (op)
+                    {
+                        case "=":
+                        case "==":
+                            return string.Format("<{0}>{1}</{0}>", tag, values);
+                        case "!=":
+                            return string.Format("<not><{0}>{1}</{0}></not>", tag, values);
+                        default:
+                            return string.Format("<{0} op='{1}'>{2}</{0}>", tag, op, values);
+                    }
+
+                case "name":
+                case "test":
+                case "fullname":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = Peek().Kind == TokenKind.String
+                        ? GetString()
+                        : GetWord();
+                    tag = lhs == "fullname" ? "test" : lhs;
```

Comment:
Yes, name is what we generate. So is test/fullname, if you remember that fixtures may be parameterized and generic as well. I'm in the same quandary as you and I'm drawing the same blank. 

The --test option has always worked for namespaces and test == namespace still works as well. Maybe having a namespace keyword would make that clearer.

If class works both ways, maybe test should do it as well.

Here's an idea. Maybe we separate the keyword list from the filters we generate. A namespace name could generate <test>My.Name.Space</test> and it would simply work. The only downside is that if you had a fixture Space, in My.Name, it would select it. Let's think about how precise we need to be with this.

Of course, you can also deal with the problem of having to qualify names by use of =~ plus a regex. At least once I implement it. :-)